feat: add `blas/ext/base/wasm/dasumpw`

aayush0325 opened the pull request:
Resolves #5509

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `blas/ext/base/wasm/dasumpw`

## Related Issues

> Does this pull request have any related issues?

Resolves #5509

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add initial wasm implementation for dasumpw

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
fix: fix bug

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
test: add tests for wasm/dasumpw

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

aayush0325 committed:
bench: add benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
feat: add examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
fix: fix package.json

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
docs: add repl.txt

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: add ts declaration file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
docs: add README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

aayush0325 committed:
chore: self review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+	* var xptr = 0;
+	*
+	* // Write vector values to module memory:
+	* mod.write( xptr, oneTo( N, dtype ) );
+	*
+	* // Perform computation:
+	* var out = mod.main( N, 5.0, xptr, 1 );
+	* // returns 21.0
+	*/
```
fix

aayush0325 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+	* // Create a BLAS routine:
+	* var mod = dasumpw.Module( mem );
+	* // returns <Module>
+	*
+	* // Initialize the routine:
+	* mod.initializeSync();
+	*
+	* // Define a vector data type:
+	* var dtype = 'float64';
+	*
+	* // Specify a vector length:
+	* var N = 3;
+	*
+	* // Define a pointer (i.e., byte offset) to the first vector element:
+	* var xptr = 0;
+	*
+	* // Write vector values to module memory:
+	* mod.write( xptr, oneTo( N, dtype ) );
+	*
+	* // Perform computation:
+	* var out = mod.main( N, 5.0, xptr, 1 );
+	* // returns 21.0
+	*/
```
fix

aayush0325 reviewed (commented):
self review

aayush0325 committed:
chore: self review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented:
ready for review @kgryte!

aayush0325 commented:
things worth documenting in further issues:

- `binary.browser.js` is an auto generated file which is made from `scripts/*` the copyright year in `scripts/template.txt` should be updated if needed.
- the `manifest.json` file of the upstream package should have a `"wasm": true` field, for an example see [blas/ext/base/dapxsumkbn/manifest.json](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/blas/ext/base/dapxsumkbn/manifest.json). if it's not there, it should be updated accordingly as done in this PR
- `package.json` should have the field given below for the compilation to work, you will not even see an error message after running the compilation command if it isn't the case

```json
"__stdlib__": {
    "wasm": true
  }
```

aayush0325 committed:
fix: add test.ts file and fix types

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented:
@aayush0325 Would you mind updating the build script and template based on the recent changes for the `wasm` packages? I updated the script to auto-insert the copyright year.

aayush0325 committed:
refactor: update build script

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented:
updated! @kgryte

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/README.md:
```diff
@@ -0,0 +1,314 @@
+## Notes
+
+-   If `N <= 0`, both `main` and `ndarray` methods return `0.0`.
+-   This package implements routines using WebAssembly. When provided arrays which are not allocated on a `dasumpw` module memory instance, data must be explicitly copied to module memory prior to computation. Data movement may entail a performance cost, and, thus, if you are using arrays external to module memory, you should prefer using [`@stdlib/blas/base/dasumpw`][@stdlib/blas/ext/base/dasump ...
```
```suggestion
-   This package implements routines using WebAssembly. When provided arrays which are not allocated on a `dasumpw` module memory instance, data must be explicitly copied to module memory prior to computation. Data movement may entail a performance cost, and, thus, if you are using arrays external to module memory, you should prefer using [`@stdlib/blas/ext/base/dasumpw`][@stdlib/blas/ext/base/dasumpw]. However, if working with arrays which are allocated and explicitly managed on module memory, you can achieve better performance when compared to the pure JavaScript implementations found in [`@stdlib/blas/ext/base/dasumpw`][@stdlib/blas/ext/base/dasumpw]. Beware that such performance gains may come at the cost of additional complexity when having to perform manual memory management. Choosing between implementations depends heavily on the particular needs and constraints of your application, with no one choice universally better than the other.
```

kgryte committed:
docs: fix package names

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/repl.txt:
```diff
@@ -0,0 +1,500 @@
+
+
+{{alias}}.Module.prototype.main( N, xp, sx )
+    Adds a scalar constant to each double-precision floating-point strided
+    array element and computes the sum using an improved Kahan–Babuška
+    algorithm.
```
```suggestion
    array element and computes the sum using pairwise summation.
```

kgryte replied:
```suggestion
    Computes the sum of absolute values (L1 norm) of double-precision
    floating-point strided array elements using pairwise summation.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/repl.txt:
```diff
@@ -0,0 +1,500 @@
+
+
+{{alias}}.Module.prototype.ndarray( N, xp, sx, ox )
+    Adds a scalar constant to each double-precision floating-point strided
+    array element and computes the sum using an improved Kahan–Babuška algorithm
+    and alternative indexing semantics.
```
```suggestion
    Computes the sum of absolute values (L1 norm) of double-precision
    floating-point strided array elements using pairwise summation and
    alternative indexing semantics.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+}
+
+/**
+* Adds a scalar constant to each double-precision floating-point strided array element and computes the sum using an improved Kahan–Babuška algorithm.
```
```suggestion
* Computes the sum of absolute values (L1 norm) of double-precision floating-point strided array elements using pairwise summation.
```

Beware of copy-paste errors.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.main( ( x: number ): number => x, x, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method is provided a second argument which is not a number...
```
This test is wrong.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.main( x.length, ( x: number ): number => x, x, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method is provided a third argument which is not a Float64Array...
```
```suggestion
// The compiler throws an error if the `main` method is provided a second argument which is not a Float64Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.main( x.length, ( x: number ): number => x, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method is provided a fourth argument which is not a number...
```
```suggestion
// The compiler throws an error if the `main` method is provided a third argument which is not a number...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.main( ( x: number ): number => x, x, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method is provided a second argument which is not a number...
+{
+	const x = new Float64Array( 10 );
+
+	dasumpw.main( x.length, '10', x, 1 ); // $ExpectError
+	dasumpw.main( x.length, true, x, 1 ); // $ExpectError
+	dasumpw.main( x.length, false, x, 1 ); // $ExpectError
+	dasumpw.main( x.length, null, x, 1 ); // $ExpectError
+	dasumpw.main( x.length, undefined, x, 1 ); // $ExpectError
+	dasumpw.main( x.length, [], x, 1 ); // $ExpectError
+	dasumpw.main( x.length, {}, x, 1 ); // $ExpectError
+	dasumpw.main( x.length, ( x: number ): number => x, x, 1 ); // $ExpectError
+}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.ndarray( ( x: number ): number => x, x, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `ndarray` method is provided a second argument which is not a number...
+{
+	const x = new Float64Array( 10 );
+
+	dasumpw.ndarray( x.length, '10', x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, true, x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, false, x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, null, x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, undefined, x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, [], x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, {}, x, 1, 0 ); // $ExpectError
+	dasumpw.ndarray( x.length, ( x: number ): number => x, x, 1, 0 ); // $ExpectError
+}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.ndarray( x.length, ( x: number ): number => x, x, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `ndarray` method is provided a third argument which is not a Float64Array...
```
```suggestion
// The compiler throws an error if the `ndarray` method is provided a second argument which is not a Float64Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.ndarray( x.length, ( x: number ): number => x, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `ndarray` method is provided a fourth argument which is not a number...
```
```suggestion
// The compiler throws an error if the `ndarray` method is provided a third argument which is not a number...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	dasumpw.ndarray( x.length, x, ( x: number ): number => x, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `ndarray` method is provided a fifth argument which is not a number...
```
```suggestion
// The compiler throws an error if the `ndarray` method is provided a fourth argument which is not a number...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/docs/types/test.ts:
```diff
@@ -0,0 +1,409 @@
+	mod.main( ( x: number ): number => x, 0, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method of a module instance is provided a second argument which is not a number...
```
Same type of mistake. Please go back through this file and remove the various copy-paste bugs.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/lib/routine.js:
```diff
@@ -0,0 +1,166 @@
+inherits( Routine, Module );
+
+/**
+* Adds a scalar constant to each double-precision floating-point strided array element and computes the sum using an improved Kahan–Babuška algorithm.
```
Copy-paste mistake.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/lib/routine.js:
```diff
@@ -0,0 +1,166 @@
+});
+
+/**
+* Adds a scalar constant to each double-precision floating-point strided array element and computes the sum using an improved Kahan–Babuška algorithm and alternative indexing semantics.
```
Copy-paste mistake.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/test/test.module.main.js:
```diff
@@ -0,0 +1,202 @@
+	t.end();
+});
+
+tape( 'a module instance has a `main` method that the sum of absolute values of all strided array elements', function test( t ) {
```
```suggestion
tape( 'a module instance has a `main` method that computes the sum of absolute values of strided array elements', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,233 @@
+	t.end();
+});
+
+tape( 'a module instance has a `ndarray` method that the sum of absolute values of all strided array elements', function test( t ) {
```
```suggestion
tape( 'a module instance has a `ndarray` method that computes the sum of absolute values of strided array elements', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,233 @@
+	t.end();
+});
+
+tape( 'if provided an `stride` parameter equal to `0`, a module instance has a `ndarray` method which returns the absolute value of the first element repeated N times', function test( t ) {
```
```suggestion
tape( 'if provided an `stride` parameter equal to `0`, a module instance has a `ndarray` method which returns the sum of the absolute value of the first element repeated N times', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/test/test.module.main.js:
```diff
@@ -0,0 +1,202 @@
+	t.end();
+});
+
+tape( 'if provided an `stride` parameter equal to `0`, a module instance has a `main` method which returns the absolute value of the first element repeated N times', function test( t ) {
```
```suggestion
tape( 'if provided an `stride` parameter equal to `0`, a module instance has a `main` method which returns sum of the absolute value of the first element repeated N times', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/test/test.ndarray.js:
```diff
@@ -0,0 +1,211 @@
+	t.end();
+});
+
+tape( 'if provided an `N` parameter equal to `1`, the function returns the first element', function test( t ) {
```
```suggestion
tape( 'if provided an `N` parameter equal to `1`, the function returns the absolute value of the first element', function test( t ) {
```

kgryte reviewed (changes requested):
This PR has a number of copy-paste mistakes which need to be addressed.

aayush0325 committed:
chore: code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into blas-wasm-dasumpw

aayush0325 committed:
fix: fix tests in test.ts

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: update test descriptions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dasumpw/test/test.ndarray.js:
```diff
@@ -0,0 +1,211 @@
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of the first element repeated N times', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of the absolute value of the first element repeated N times', function test( t ) {
```

kgryte committed:
test: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
